src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/LoginHandler.cs

```diff
@@ -0,0 +1,636 @@
+                    }
+                    else
+                    {
+                        stream.Write(outSSPIBuff.AsSpan(0, (int)outSSPILength));
+                    }
+                }
+                await stream.WriteStringAsync(context.AttachedDbFileName, isAsync, ct).ConfigureAwait(false);
+                if (!context.UseSspi && !(context.Features.FederatedAuthenticationInfoRequested || context.Features.FederatedAuthenticationRequested))
+                {
+                    if (context.PasswordChangeRequest?.NewSecurePassword != null)
+                    {
+                        // TODO : implement saving secure string.
+                        throw new NotImplementedException();
+                        // _physicalStateObj.WriteSecureString(rec.newSecurePassword);
+                    }
+                    else
+                    {
+                        if (isAsync)
+                        {
+                            await stream.WriteAsync(encryptedChangePassword.AsMemory(0, encryptedChangePasswordLengthInBytes), ct).ConfigureAwait(false);
+                        }
+                        else
+                        {
+                            stream.Write(encryptedChangePassword.AsSpan(0, encryptedChangePasswordLengthInBytes));
+                        }
+                    }
+                }
+
+                await SendFeatureExtensionData(context, 
+                    requestedFeatures, 
+                    recoverySessionData, 
+                    fedAuthFeatureExtensionData, 
+                    isAsync, 
+                    ct).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                if (ADP.IsCatchableExceptionType(e))
+                {
+                    // Reset the buffer if there was an exception.
+                    context.TdsStream.Reset();
+                }
+
+                throw;
+            }
+        }
+
+        private async ValueTask SendFeatureExtensionData(LoginHandlerContext context,
+                                        TdsEnums.FeatureExtension requestedFeatures,
+                                        SessionData recoverySessionData,
+                                        FederatedAuthenticationFeatureExtensionData fedAuthFeatureExtensionData,
+                                        bool isAsync,
+                                        CancellationToken ct)
+        {
+            if (context.UseFeatureExt)
+            {
+                if (_sessionRecoveryFeature.ShouldUseFeature(requestedFeatures))
+                {
+                    await _sessionRecoveryFeature.WriteFeatureData(context, isAsync, ct).ConfigureAwait(false);
+                }
+                if (_fedAuthFeature.ShouldUseFeature(requestedFeatures))
+                {
+                    SqlClientEventSource.Log.TryTraceEvent("<sc.TdsParser.TdsLogin|SEC> Sending federated authentication feature request");
```

Comment:
Trace events can be made internal to WriteFeatureData too based on whether feature is requested and is being written.

Reply:
Removing traceevents. That requires additional thinking.